libraries/libstratosphere/source/settings/impl/settings_key_value_store.cpp

```diff
@@ -0,0 +1,1629 @@
+        FreeToHeap(out->map_key, out->key_size);
+
+        /* Allocate the new map key. */
+        char *buffer = static_cast<char *>(AllocateFromHeap(map_key_size));
+        AMS_ASSERT(buffer != nullptr);
+
+        /* Copy the new map key to the buffer. */
+        std::memcpy(buffer, map_key.GetString(), map_key_size);
+
+        /* Set the output map key. */
+        out->key_size = map_key_size;
+        out->map_key = buffer;
+
+        return ResultSuccess();
+    }
+    
+    Result DestroyKeyValueStoreKeyIterator(KeyValueStoreKeyIterator *out) {
+        /* Check preconditions. */
+        AMS_ASSERT(out != nullptr);
+        AMS_ASSERT(out->name_size > 0);
+        AMS_ASSERT(out->name_size < out->key_size);
+        AMS_ASSERT(out->map_key != nullptr);
+
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Free the key to the heap. */
+        FreeToHeap(out->map_key, out->key_size);
+
+        /* Reset the name and key. */
+        out->name_size = 0;
+        out->key_size = 0;
+        out->map_key = nullptr;
+        return ResultSuccess();
+    }
+
+    Result GetKeyValueStoreItemCountForDebug(size_t *out_count) {
+        /* Check preconditions. */
+        AMS_ASSERT(out != nullptr);
+
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Get the key value store map. */
+        Map *map = nullptr;
+        R_TRY(GetKeyValueStoreMap(std::addressof(map)));
+        AMS_ASSERT(map != nullptr);
+
+        /* Output the item count. */
+        *out_count = map->size();
+        return ResultSuccess();
+    }
+    
+    Result GetKeyValueStoreItemForDebug(size_t *out_count, KeyValueStoreItemForDebug * const out_items, size_t out_items_count) {
+        /* Check preconditions. */
+        AMS_ASSERT(out_count != nullptr);
+        AMS_ASSERT(out_items != nullptr);
+
+        /* Acquire exclusive access to global state. */
+        std::scoped_lock lk(g_key_value_store_mutex);
+
+        /* Get the key value store map. */
+        Map *map = nullptr;
+        R_TRY(GetKeyValueStoreMap(std::addressof(map)));
+        AMS_ASSERT(map != nullptr);
+
+        /* Define the count variable. */
+        u64 count = 0;
+
+        /* Iterate through each value in the map and output kvs items. */
+        for (auto it = map->begin(); it != map->end(); ++it) {
+            /* Get the current map value. */
+            const MapValue &map_value = it->second;
+
+            /* Break if the count exceeds the output items count. */
+            if (count >= out_items_count) {
+                break;
+            }
+
+            /* Get the current item. */ 
+            KeyValueStoreItemForDebug &item = out_items[count++];
+
+            /* Copy the map key and value to the item. */
+            item.key                = const_cast<char *>(it->first.GetString());
```

Comment:
wtf why is this a const_cast? This shouldn't need a const_cast.

Wait I forgot I told you to make the .key field const char * up above, so just fix that there.